File: android/src/com/mku/salmon/vault/provider/SalmonFileProvider.java
package com.mku.salmon.vault.provider;

import static android.provider.DocumentsContract.Root;

import android.database.Cursor;
import android.database.MatrixCursor;
import android.os.CancellationSignal;
import android.os.Handler;
import android.os.Parcel;
import android.os.ParcelFileDescriptor;
import android.provider.DocumentsContract;
import android.provider.DocumentsProvider;
import android.webkit.MimeTypeMap;

import androidx.annotation.Nullable;

import com.mku.android.salmonfs.drive.AndroidDrive;
import com.mku.convert.BitConverter;
import com.mku.fs.drive.utils.FileUtils;
import com.mku.fs.file.File;
import com.mku.fs.file.IFile;
import com.mku.salmon.Generator;
import com.mku.salmon.vault.android.R;
import com.mku.salmon.vault.main.SalmonApplication;
import com.mku.salmon.vault.model.SalmonVaultManager;
import com.mku.salmon.vault.services.AndroidSettingsService;
import com.mku.salmon.vault.services.ISettingsService;
import com.mku.salmon.vault.services.ServiceLocator;
import com.mku.salmonfs.drive.AesDrive;
import com.mku.salmonfs.file.AesFile;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;

public class SalmonFileProvider extends DocumentsProvider {
    public static final long MAX_FILE_SIZE_TO_SHARE = 128 * 1024 * 1024;
    public static final long MEDIUM_FILE_SIZE_TO_SHARE = 10 * 1024 * 1024;
    private static String rootPath = "/";

    private static HashMap<String, Boolean> authorizedApps = new HashMap<>();
    private static String[] rootProjection = new String[]{
            DocumentsContract.Root.COLUMN_ROOT_ID,
            DocumentsContract.Root.COLUMN_DOCUMENT_ID,
            DocumentsContract.Root.COLUMN_ICON,
            DocumentsContract.Root.COLUMN_TITLE,
            DocumentsContract.Root.COLUMN_FLAGS};

    private static String[] documentProjection = new String[]{
            DocumentsContract.Document.COLUMN_DOCUMENT_ID,
            DocumentsContract.Document.COLUMN_DISPLAY_NAME,
            DocumentsContract.Document.COLUMN_MIME_TYPE,
            DocumentsContract.Document.COLUMN_FLAGS,
            DocumentsContract.Document.COLUMN_SIZE,
            DocumentsContract.Document.COLUMN_LAST_MODIFIED,
            DocumentsContract.Document.COLUMN_ICON};

    private boolean hasServices;

    public static void authorizeApp(String packageName) {
        if (packageName != null && authorizedApps.containsKey(packageName))
            authorizedApps.put(packageName, true);
    }

    public static java.io.File createSharedFile(AesFile salmonFile) throws Exception {
        java.io.File sharedFile = ((AndroidDrive) SalmonVaultManager.getInstance().getDrive())
                .copyToSharedFolder(salmonFile);
        byte[] rand = Generator.getSecureRandomBytes(32);
        java.io.File dir = new java.io.File(sharedFile.getParentFile(), BitConverter.toHex(rand));
        if(!dir.mkdir())
            throw new RuntimeException("Could not create dir");
        java.io.File nFile = new java.io.File(dir, sharedFile.getName());
        if(!sharedFile.renameTo(nFile))
            throw new RuntimeException("Could not rename file");
        return nFile;
    }

    protected void setupServices() {
        if (hasServices)
            return;
        ServiceLocator.getInstance().register(ISettingsService.class, new AndroidSettingsService());
        hasServices = true;
    }

    @Override
    public Cursor queryRoots(String[] projection) {
        final MatrixCursor result = new MatrixCursor(rootProjection);
        final MatrixCursor.RowBuilder row = result.newRow();
        getRoot(row);
        return result;
    }

    private void getRoot(MatrixCursor.RowBuilder row) {
        row.add(Root.COLUMN_ROOT_ID, rootPath);
        row.add(Root.COLUMN_DOCUMENT_ID, rootPath);
        row.add(Root.COLUMN_TITLE, "Salmon Files");
        row.add(Root.COLUMN_ICON, R.drawable.logo_128x128);
    }

    private void getRootDocument(MatrixCursor.RowBuilder row) {
        AesDrive drive = SalmonVaultManager.getInstance().getDrive();
        row.add(DocumentsContract.Document.COLUMN_DOCUMENT_ID, rootPath);
        row.add(DocumentsContract.Document.COLUMN_DISPLAY_NAME, rootPath);
        row.add(DocumentsContract.Document.COLUMN_MIME_TYPE,
                DocumentsContract.Document.MIME_TYPE_DIR);
        row.add(DocumentsContract.Document.COLUMN_FLAGS, 0);
        row.add(DocumentsContract.Document.COLUMN_SIZE, 0);
        row.add(DocumentsContract.Document.COLUMN_LAST_MODIFIED, System.currentTimeMillis());
    }

    private void getErrorDocument(MatrixCursor.RowBuilder row, String text) {
        row.add(DocumentsContract.Document.COLUMN_DOCUMENT_ID, "0");
        row.add(DocumentsContract.Document.COLUMN_DISPLAY_NAME, text);
        row.add(DocumentsContract.Document.COLUMN_MIME_TYPE, "application/octetstream");
        row.add(DocumentsContract.Document.COLUMN_FLAGS, 0);
        row.add(DocumentsContract.Document.COLUMN_SIZE, 0);
        row.add(DocumentsContract.Document.COLUMN_LAST_MODIFIED, System.currentTimeMillis());
        row.add(DocumentsContract.Document.COLUMN_ICON, R.drawable.info_small);
    }

    private void getDocument(MatrixCursor.RowBuilder row, AesFile file) throws IOException {
        row.add(DocumentsContract.Document.COLUMN_DOCUMENT_ID, file.getPath());
        row.add(DocumentsContract.Document.COLUMN_DISPLAY_NAME, file.getName());
        if (file.isDirectory())
            row.add(DocumentsContract.Document.COLUMN_MIME_TYPE, DocumentsContract.Document.MIME_TYPE_DIR);
        else {
            String ext = FileUtils.getExtensionFromFileName(file.getName()).toLowerCase();
            String mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(ext);
            if (mimeType == null)
                mimeType = "application/octetstream";
            row.add(DocumentsContract.Document.COLUMN_MIME_TYPE, mimeType);
        }
        int flags = 0;
        if (file.isFile())
            flags |= DocumentsContract.Document.FLAG_SUPPORTS_WRITE;
        row.add(DocumentsContract.Document.COLUMN_FLAGS, flags);
        row.add(DocumentsContract.Document.COLUMN_SIZE, file.getRealFile().getLength());
        row.add(DocumentsContract.Document.COLUMN_LAST_MODIFIED, file.getLastDateModified());
        if (file.isDirectory())
            row.add(DocumentsContract.Document.COLUMN_ICON, R.drawable.folder);
        else
            row.add(DocumentsContract.Document.COLUMN_ICON, R.drawable.file_item_small);
    }

    @Override
    public Cursor queryDocument(String documentId, String[] projection) {
        setupServices();
        final MatrixCursor result = new MatrixCursor(documentProjection);
        AesDrive drive = getManager().getDrive();
        final MatrixCursor.RowBuilder row = result.newRow();
		if (drive == null) {
            getErrorDocument(row, "No opened vaults");
            return result;
        }

        if (documentId.equals(rootPath))
            getRootDocument(row);
        else {
            if (!checkAppAuthorized()) {
                return result;
            }
            try {
                AesFile file = parsePath(documentId);
                getDocument(row, file);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
        return result;
    }

    @Override
    public Cursor queryChildDocuments(String parentDocumentId, String[] projection, String sortOrder) throws FileNotFoundException {
        final MatrixCursor result = new MatrixCursor(documentProjection);
        if (getManager().getDrive() == null) {
            MatrixCursor.RowBuilder row = result.newRow();
            getErrorDocument(row, "No opened vaults");
            return result;
        }
        if (!checkAppAuthorized()) {
            MatrixCursor.RowBuilder row = result.newRow();
            getErrorDocument(row, "App not authorized");
            return result;
        }
        try {
            AesFile dir = parsePath(parentDocumentId);
            AesFile[] files = dir.listFiles();
            for (AesFile file : files) {
                MatrixCursor.RowBuilder row = result.newRow();
                getDocument(row, file);
            }
        } catch (IOException e) {
            throw new FileNotFoundException(e.toString());
        }
        return result;
    }

    private AesFile parsePath(String documentId) throws IOException {
        String[] parts = documentId.split("/");
        AesFile file = getManager().getDrive().getRoot();
        for (int i = 1; i < parts.length; i++) {
            file = file.getChild(parts[i]);
        }
        return file;
    }

    @Override
    public ParcelFileDescriptor openDocument(String documentId, String mode,
                                             @Nullable CancellationSignal signal)
            throws FileNotFoundException {
        if (!checkAppAuthorized()) {
            throw new FileNotFoundException("App not authorized, use Salmon Vault app to allow access");
        }
        // TODO: check the CancellationSignal periodically
        AesFile salmonFile;
        java.io.File sharedFile;
        String filename;
        final int accessMode = ParcelFileDescriptor.parseMode(mode);
        try {
            salmonFile = parsePath(documentId);
            if (salmonFile.getLength() > MAX_FILE_SIZE_TO_SHARE) {
                throw new RuntimeException(SalmonApplication.getInstance().getString(R.string.FileSizeTooLarge));
            }
            filename = salmonFile.getName();
            sharedFile = createSharedFile(salmonFile);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }

        ParcelFileDescriptor descriptor;
        final boolean isWrite = (mode.indexOf('w') != -1);
        if (isWrite) {
            try {
                Handler handler = new Handler(getContext().getMainLooper());
                AesFile file = salmonFile;
                IFile importFile = new File(sharedFile.getPath());
                String finalFilename = filename;
                descriptor = ParcelFileDescriptor.open(sharedFile, accessMode, handler,
                        e -> {
                            AesFile parentDir = file.getParent();
                            getManager().importFiles(new IFile[]{importFile}, parentDir, false,
                                    (AesFile[] importedAesFiles) ->
                                    {
                                        try {
                                            if (!importedAesFiles[0].exists())
                                                return;
                                            importedAesFiles[0].rename(finalFilename);
                                            file.delete();
                                            if (getManager().getDrive() != null)
                                                getManager().refresh();
                                        } catch (Exception ex) {
                                            ex.printStackTrace();
                                            throw new RuntimeException("Could not reimport file: " + ex.getMessage());
                                        }
                                    });
                        });
            } catch (IOException e) {
                throw new FileNotFoundException("Could not reimport shared file: " + e.getMessage());
            }
        } else {
            descriptor = ParcelFileDescriptor.open(sharedFile, accessMode);
        }
        return descriptor;
    }

    private boolean checkAppAuthorized() {
        String callingPackageName = getCallingPackage();
        boolean allowed = false;
        if (!authorizedApps.containsKey(callingPackageName)) {
            authorizedApps.put(callingPackageName, false);
        } else {
            Boolean res = authorizedApps.get(callingPackageName);
            allowed = res == null ? false : res;
        }
        return allowed;
    }

    public static ArrayList<String> getApps(boolean authorized) {
        ArrayList<String> list = new ArrayList<>();
        for (String packageName : authorizedApps.keySet()) {
            Boolean allowed = authorizedApps.get(packageName);
            if (packageName != null && allowed != null && authorized == allowed)
                list.add(packageName);
        }
        return list;
    }

    @Override
    public boolean onCreate() {
        return false;
    }

    protected SalmonVaultManager getManager() {
        return SalmonVaultManager.getInstance();
    }
}

File: dotnet-android/src/Provider/SalmonFileProvider.cs
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
using Mku.Utils;
using Mku.Salmon;
using BitConverter = Mku.Convert.BitConverter;
using System;
using System.Collections.Generic;
using Java.Lang;
using Android.Provider;
using Mku.Android.Salmon.Drive;
using Salmon.Vault.Model;
using Salmon.Vault.Services;
using Android.Database;
using static Android.Provider.DocumentsContract;
using Salmon.Vault.DotNetAndroid;
using System.IO;
using Android.Webkit;
using Mku.File;
using static Android.OS.ParcelFileDescriptor;
using Salmon.Vault.Main;

namespace Salmon.Vault.Provider;

public class SalmonFileProvider : DocumentsProvider
{
    public static readonly long MAX_FILE_SIZE_TO_SHARE = 50 * 1024 * 1024;
    public static readonly long MEDIUM_FILE_SIZE_TO_SHARE = 10 * 1024 * 1024;
    private static string rootPath = "/";

    private static Dictionary<string, bool> authorizedApps = new Dictionary<string, bool>();
    private static string[] rootProjection = new string[]{
            DocumentsContract.Root.ColumnRootId,
            DocumentsContract.Root.ColumnDocumentId,
            DocumentsContract.Root.ColumnIcon,
            DocumentsContract.Root.ColumnTitle,
            DocumentsContract.Root.ColumnFlags};

    private static string[] documentProjection = new string[]{
            DocumentsContract.Document.ColumnDocumentId,
            DocumentsContract.Document.ColumnDisplayName,
            DocumentsContract.Document.ColumnMimeType,
            DocumentsContract.Document.ColumnFlags,
            DocumentsContract.Document.ColumnSize,
            DocumentsContract.Document.ColumnLastModified,
            DocumentsContract.Document.ColumnIcon};

    private bool hasServices;

    public static void authorizeApp(string packageName)
    {
        if (packageName != null && authorizedApps.ContainsKey(packageName))
            authorizedApps[packageName] = true;
    }

    public static Java.IO.File CreateSharedFile(SalmonFile salmonFile)
    {
        Java.IO.File sharedFile = ((AndroidDrive)SalmonVaultManager.Instance.Drive)
                .CopyToSharedFolder(salmonFile);
        byte[] rand = SalmonGenerator.GetSecureRandomBytes(32);
        Java.IO.File dir = new Java.IO.File(sharedFile.ParentFile, BitConverter.ToHex(rand));
        if (!dir.Mkdir())
            throw new RuntimeException("Could not create dir");
        Java.IO.File nFile = new Java.IO.File(dir, sharedFile.Name);
        if (!sharedFile.RenameTo(nFile))
            throw new RuntimeException("Could not rename file");
        return nFile;
    }

    protected void setupServices()
    {
        if (hasServices)
            return;
        ServiceLocator.GetInstance().Register(typeof(ISettingsService), new AndroidSettingsService());
        hasServices = true;
    }

    public override ICursor QueryRoots(string[] projection)
    {
        MatrixCursor result = new MatrixCursor(rootProjection);
        MatrixCursor.RowBuilder row = result.NewRow();
        getRoot(row);
        return result;
    }

    private void getRoot(MatrixCursor.RowBuilder row)
    {
        row.Add(Root.ColumnRootId, rootPath);
        row.Add(Root.ColumnDocumentId, rootPath);
        row.Add(Root.ColumnTitle, "Salmon Files");
        row.Add(Root.ColumnIcon, Resource.Drawable.logo_128x128);
    }

    private void getRootDocument(MatrixCursor.RowBuilder row)
    {
        SalmonDrive drive = SalmonVaultManager.Instance.Drive;
        row.Add(DocumentsContract.Document.ColumnDocumentId, rootPath);
        row.Add(DocumentsContract.Document.ColumnDisplayName, rootPath);
        row.Add(DocumentsContract.Document.ColumnMimeType,
                DocumentsContract.Document.MimeTypeDir);
        row.Add(DocumentsContract.Document.ColumnFlags, 0);
        row.Add(DocumentsContract.Document.ColumnSize, 0);
        row.Add(DocumentsContract.Document.ColumnLastModified, DateTime.Now.Millisecond);
    }

    private void getErrorDocument(MatrixCursor.RowBuilder row, string text)
    {
        row.Add(DocumentsContract.Document.ColumnDocumentId, "0");
        row.Add(DocumentsContract.Document.ColumnDisplayName, text);
        row.Add(DocumentsContract.Document.ColumnMimeType, "application/octetstream");
        row.Add(DocumentsContract.Document.ColumnFlags, 0);
        row.Add(DocumentsContract.Document.ColumnSize, 0);
        row.Add(DocumentsContract.Document.ColumnLastModified, DateTime.Now.Millisecond);
        row.Add(DocumentsContract.Document.ColumnIcon, Resource.Drawable.info_small);
    }

    private void getDocument(MatrixCursor.RowBuilder row, SalmonFile file)
    {
        row.Add(DocumentsContract.Document.ColumnDocumentId, file.Path);
        row.Add(DocumentsContract.Document.ColumnDisplayName, file.BaseName);
        if (file.IsDirectory)
            row.Add(DocumentsContract.Document.ColumnMimeType, DocumentsContract.Document.MimeTypeDir);
        else
        {
            string ext = FileUtils.GetExtensionFromFileName(file.BaseName).ToLower();
            string mimeType = MimeTypeMap.Singleton.GetMimeTypeFromExtension(ext);
            if (mimeType == null)
                mimeType = "application/octetstream";
            row.Add(DocumentsContract.Document.ColumnMimeType, mimeType);
        }
        int? flags = null;
        if (file.IsFile)
            flags = (int?)DocumentContractFlags.SupportsWrite;
        row.Add(DocumentsContract.Document.ColumnFlags, flags);
        row.Add(DocumentsContract.Document.ColumnSize, file.RealFile.Length);
        row.Add(DocumentsContract.Document.ColumnLastModified, file.LastDateTimeModified);
        if (file.IsDirectory)
            row.Add(DocumentsContract.Document.ColumnIcon, Resource.Drawable.folder);
        else
            row.Add(DocumentsContract.Document.ColumnIcon, Resource.Drawable.file_item_small);
    }

    public override ICursor QueryDocument(string documentId, string[] projection)
    {
        setupServices();
        MatrixCursor result = new MatrixCursor(documentProjection);
        SalmonDrive drive = GetManager().Drive;
        MatrixCursor.RowBuilder row = result.NewRow();
        if (drive == null)
        {
            getErrorDocument(row, "No opened vaults");
            return result;
        }

        if (documentId.Equals(rootPath))
            getRootDocument(row);
        else
        {
            if (!CheckAppAuthorized())
            {
                return result;
            }
            try
            {
                SalmonFile file = ParsePath(documentId);
                getDocument(row, file);
            }
            catch (IOException e)
            {
                throw new RuntimeException(e.Message);
            }
        }
        return result;
    }

    public override ICursor QueryChildDocuments(string parentDocumentId, string[] projection, string sortOrder)
    {
        MatrixCursor result = new MatrixCursor(documentProjection);
        if (GetManager().Drive == null)
        {
            MatrixCursor.RowBuilder row = result.NewRow();
            getErrorDocument(row, "No opened vaults");
            return result;
        }
        if (!CheckAppAuthorized())
        {
            MatrixCursor.RowBuilder row = result.NewRow();
            getErrorDocument(row, "App not authorized");
            return result;
        }
        try
        {
            SalmonFile dir = ParsePath(parentDocumentId);
            SalmonFile[] files = dir.ListFiles();
            foreach (SalmonFile file in files)
            {
                MatrixCursor.RowBuilder row = result.NewRow();
                getDocument(row, file);
            }
        }
        catch (IOException e)
        {
            throw new FileNotFoundException(e.ToString());
        }
        return result;
    }

    private SalmonFile ParsePath(string documentId)
    {
        string[] parts = documentId.Split("/");
        SalmonFile file = GetManager().Drive.Root;
        for (int i = 1; i < parts.Length; i++)
        {
            if (parts[i] == "")
                continue;
            file = file.GetChild(parts[i]);
        }
        return file;
    }

    class OnCloseListener : Java.Lang.Object, IOnCloseListener
    {
        SalmonFile file;
        IRealFile importFile;
        SalmonFileProvider provider;
        string filename;

        public OnCloseListener(SalmonFileProvider provider, SalmonFile file, IRealFile importFile, string filename)
        {
            this.file = file;
            this.importFile = importFile;
            this.provider = provider;
            this.filename = filename;
        }

        public void OnClose(Java.IO.IOException? ex)
        {
            SalmonFile parentDir = file.Parent;
            provider.GetManager().ImportFiles(new IRealFile[] { importFile }, parentDir, false,
                (SalmonFile[] importedSalmonFiles) =>
                {
                    try
                    {
                        if (!importedSalmonFiles[0].Exists)
                            return;
                        importedSalmonFiles[0].Rename(filename);
                        file.Delete();
                        if (provider.GetManager().Drive != null)
                            provider.GetManager().Refresh();
                    }
                    catch (Java.Lang.Exception ex)
                    {
                        ex.PrintStackTrace();
                        throw new RuntimeException("Could not reimport file: " + ex.Message);
                    }
                });
        }
    }

    public override Android.OS.ParcelFileDescriptor OpenDocument(string documentId, string mode, Android.OS.CancellationSignal signal)
    {
        if (!CheckAppAuthorized())
        {
            // App not authorized
            return null;
        }
        // TODO: check the CancellationSignal periodically
        SalmonFile salmonFile;
        Java.IO.File sharedFile;
        string filename;
        Android.OS.ParcelFileMode accessMode = Android.OS.ParcelFileDescriptor.ParseMode(mode);
        try
        {
            salmonFile = ParsePath(documentId);
            if (salmonFile.Size > MAX_FILE_SIZE_TO_SHARE)
            {
                throw new RuntimeException(SalmonApplication.GetInstance().GetString(Resource.String.FileSizeTooLarge));
            }
            filename = salmonFile.BaseName;
            sharedFile = CreateSharedFile(salmonFile);
        }
        catch (System.Exception e)
        {
            throw new RuntimeException(e.Message);
        }

        Android.OS.ParcelFileDescriptor descriptor;
        bool isWrite = (mode.IndexOf('w') != -1);
        if (isWrite)
        {
            try
            {
                Android.OS.Handler handler = new Android.OS.Handler(Context.MainLooper);
                SalmonFile file = salmonFile;
                IRealFile importFile = new DotNetFile(sharedFile.Path);
                descriptor = Android.OS.ParcelFileDescriptor.Open(sharedFile, accessMode, handler, new OnCloseListener(this, file, importFile, filename));
            }
            catch (IOException e)
            {
                throw new FileNotFoundException("Could not reimport shared file: " + e.Message);
            }
        }
        else
        {
            descriptor = Android.OS.ParcelFileDescriptor.Open(sharedFile, accessMode);
        }
        return descriptor;
    }

    private bool CheckAppAuthorized()
    {
        string callingPackageName = CallingPackage;
        bool allowed = false;
        if (callingPackageName == null)
            return false;
        if (!authorizedApps.ContainsKey(callingPackageName))
        {
            authorizedApps[callingPackageName] = false;
        }
        else
        {
            bool res = authorizedApps[callingPackageName];
            allowed = res == null ? false : res;
        }
        return allowed;
    }

    public static List<string> GetApps(bool authorized)
    {
        List<string> list = new List<string>();
        foreach (string packageName in authorizedApps.Keys)
        {
            bool allowed = authorizedApps[packageName];
            if (packageName != null && allowed != null && authorized == allowed)
                list.Add(packageName);
        }
        return list;
    }

    public override bool OnCreate()
    {
        return false;
    }

    protected SalmonVaultManager GetManager()
    {
        return SalmonVaultManager.Instance;
    }
}
